src/py/flwr/server/superlink/driver/driver_servicer.py

```diff
@@ -208,13 +204,47 @@ def PullServerAppInputs(
         self, request: PullServerAppInputsRequest, context: grpc.ServicerContext
     ) -> PullServerAppInputsResponse:
         """Pull ServerApp process inputs."""
-        raise NotImplementedError()
+        log(DEBUG, "DriverServicer.PullServerAppInputs")
+        state = self.state_factory.state()
+        ffs = self.ffs_factory.ffs()
+
+        with self.lock:
+            if request.HasField("run_id"):
+                run_id: Optional[int] = request.run_id
+            else:
+                run_id = state.get_pending_run_id()
+            if run_id is None:
+                return PullServerAppInputsResponse()
+
+            serverapp_ctxt = state.get_serverapp_context(run_id)
+            run = state.get_run(run_id)
+            fab = None
+            if run and run.fab_hash:
+                if result := ffs.get(run.fab_hash):
+                    fab = Fab(run.fab_hash, result[0])
+
+            if serverapp_ctxt and run and fab:
```

Comment:
what if `serverapp_ctxt` is `None`?

Reply:
oops, my bad. The ServerApp context can be `None` here, and we will just return `None` to the `flwr-serverapp` process.

Reply:
I've updated the code.